### PR TITLE
e2e: Simplify test context management

### PR DESCRIPTION
**Pull request by nirs**
Simplify the way we managed text context and prepare for using it in the entire e2e module.

- Eliminate the global contexts map and complicated logic to access the context from different modules by moving the action wrappers to the contexts.
- Rename testcontext.TestContext to test.Context
- Simplify test generation
- Add test.Context.Name
- Add per-test loger to test.Context
- Eliminate interface dependencies with the e2e/types package
- Implement the e2e/types.Context interface
- Introduce test.Context.Validate()
- Return more context from kubectl error
- Remove unneeded log argument
- Pass test context to utility functions
- Add test.Context.Namsapce()

With these changes we are ready to complete #1683.

### CODE REVIEW

**Comment by nirs** [comment]
Summary from discussion with @raghavendra-talur:

We want to have a DRTest struct keeping everything needed for a test:

```go
type DRTest struct {
    Workload ...
    Deployer ...
    Log ...
}
```

We will not implement any wrappers like Deploy/Undeploy/Protect/Unprotect/Failover/Relocate on the context for now, using functions to create sub tests. We can add this later when we have more flows.

We want to create this struct when we start a test, and pass it around instead of passing multiple arguments to all utility functions. This way we have access to everything related to a single test everywhere, and it is very easy to add or remove stuff without changing function signature everywhere.

This can also be used to eliminate the global Ctx later.

**Comment by nirs** [comment]
> Summary from discussion with @raghavendra-talur:
> 
> We want to have a DRTest struct keeping everything needed for a test:
> 
> ```go
> type DRTest struct {
>     Workload ...
>     Deployer ...
>     Log ...
> }
> ```

DRTest or Test does not works well like test.Context, so I kept the current name.

> We will not implement any wrappers like Deploy/Undeploy/Protect/Unprotect/Failover/Relocate on the context for now, using functions to create sub tests. We can add this later when we have more flows.

Tried this, it is 2 times longer, full of boilerplate and required multiple //nolint: to make the linter happy. Keeping the actions are methods on the context is much better so I kept the current change.

Also tried table based test - with method on the context is very nice:

```go
func runTestFlow(t *testing.T, ctx testcontext.TestContext) {
    t.Helper()

    if !ctx.Deployer.IsWorkloadSupported(ctx.Workload) {
        t.Skipf("Workload %s not supported by deployer %s, skip test", ctx.Workload.GetName(), ctx.Deployer.GetName())
    }   

    flow := []testcontext.Step{
        {Name: "Deploy", Func: ctx.Deploy},
        {Name: "Enable", Func: ctx.Enable},
        {Name: "Failover", Func: ctx.Failover},
        {Name: "Relocate", Func: ctx.Relocate},
        {Name: "Disable", Func: ctx.Disable},
        {Name: "Undeploy", Func: ctx.Uneploy},
    }   

    for _, step := range flow {
        if !t.Run(step.Name, step.Func) {
            t.Fatalf("%s failed", step.Name)
        }   
    }   
}    
```

But the current test is much simpler code so I kept is as is. If we will have multiple flows or much long flow, table based test is the way to keep it clean and easy to modify.

The key to keep is nice is to use methods on the context.

**Review by raghavendra-talur — COMMENTED** [review]
There is only one small typo that I would like to see fixed before merging this PR. We can make more enhancements later.

File: e2e/testcontext/testcontext.go [hunk 1]
```diff
@@ -17,46 +16,50 @@ type TestContext struct {
...
+func (c *TestContext) Deploy(t *testing.T) {
+	t.Helper()
 
-var contexts = make(map[string]TestContext)
+	if err := c.Deployer.Deploy(c.Workload); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (c *TestContext) Uneploy(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nirs) [reply e2e/testcontext/testcontext.go]: fixed

**Comment by nirs** [comment]
The e2e failure looks like #1659

**Review by raghavendra-talur — APPROVED** [review]

**Pull request merged**